nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/OidcService.java

```diff
@@ -0,0 +1,207 @@
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.nifi.web.security.oidc.StandardOidcIdentityProvider.OPEN_ID_CONNECT_SUPPORT_IS_NOT_CONFIGURED;
+
+/**
+ * OidcService is a service for managing the OpenId Connect Authorization flow.
+ */
+public class OidcService {
+
+    private OidcIdentityProvider identityProvider;
+    private Cache<CacheKey, State> stateLookupForPendingRequests; // identifier from cookie -> state value
+    private Cache<CacheKey, String> jwtLookupForCompletedRequests; // identifier from cookie -> jwt or identity (and generate jwt on retrieval)
+
+    /**
+     * Creates a new OtpService with an expiration of 5 minutes.
+     */
+    public OidcService(final OidcIdentityProvider identityProvider) {
+        this(identityProvider, 60, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Creates a new OtpService.
+     *
+     * @param duration                  The expiration duration
+     * @param units                     The expiration units
+     * @throws NullPointerException     If units is null
+     * @throws IllegalArgumentException If duration is negative
+     */
+    public OidcService(final OidcIdentityProvider identityProvider, final int duration, final TimeUnit units) {
+        this.identityProvider = identityProvider;
+        this.stateLookupForPendingRequests = CacheBuilder.newBuilder().expireAfterWrite(duration, units).build();
+        this.jwtLookupForCompletedRequests = CacheBuilder.newBuilder().expireAfterWrite(duration, units).build();
+    }
+
+    /**
+     * Returns whether OpenId Connect is enabled.
+     *
+     * @return whether OpenId Connect is enabled
+     */
+    public boolean isOidcEnabled() {
+        return identityProvider.isOidcEnabled();
+    }
+
+    /**
+     * Returns the OpenId Connect authorization endpoint.
+     *
+     * @return the authorization endpoint
+     */
+    public URI getAuthorizationEndpoint() {
+        return identityProvider.getAuthorizationEndpoint();
+    }
+
+    /**
+     * Returns the OpenId Connect scope.
+     *
+     * @return scope
+     */
+    public Scope getScope() {
+        return identityProvider.getScope();
+    }
+
+    /**
+     * Returns the OpenId Connect client id.
+     *
+     * @return client id
+     */
+    public String getClientId() {
+        return identityProvider.getClientId().getValue();
+    }
+
+    /**
+     * Initiates an OpenId Connection authorization code flow using the specified request identifier to maintain state.
+     *
+     * @param oidcRequestIdentifier request identifier
+     * @return state
+     */
+    public State createState(final String oidcRequestIdentifier) {
+        if (!isOidcEnabled()) {
+            throw new IllegalStateException(OPEN_ID_CONNECT_SUPPORT_IS_NOT_CONFIGURED);
+        }
+
+        final CacheKey oidcRequestIdentifierKey = new CacheKey(oidcRequestIdentifier);
+        final State state = new State(new BigInteger(130, new SecureRandom()).toString(32));
```

Comment:
Got it. Thanks for the additional explanation. The new comment will include these details.